### ISSUE TITLE
Error on Loading VB Editor: InvalidCastException: Unable to cast object of type 'VBClassicExtensibility.VBProjectClass' to type 'Microsoft.Vbe.Interop._VBProject'

### ISSUE THREAD

**Issue by christophocles** [rubberduck-vba/Rubberduck#3837]
When I open the VB editor with RubberDuck installed, I am getting a messagebox pop up that says:

"InvalidCastException: Unable to cast object of type 'VBClassicExtensibility.VBProjectClass' to type 'Microsoft.Vbe.Interop._VBProject'".

This messagebox comes up repeatedly whenever I click on a project name in the built-in VBIDE project explorer.  At first I thought it may be related to having locked some locked VBA projects, or projects that contain API calls, or perhaps the installation method (reference issue #3832).

I did a lot of troubleshooting today on two different computers.  On one computer I was unable to reproduce this error at all, and on the other it seems to randomly come up.  I now believe that the difference is having multiple different versions of office programs installed on the computer that is getting the errors.  On the "good" computer I only ever had Office 2013 installed.  On the "bad" one I have Office 2013, and used to have Access 97, 2010, and 2013.  I believe the Microsoft.VBE.Interop assemblies got messed up somehow through the series of installations on this computer, and the issue is surfacing in these InvalidTypeCast exceptions, similar to issue #3033 

I will try reinstalling MS Office on this PC and see if it helps.  But perhaps this exception could be caught and handled more gracefully so that these annoying messageboxes don't keep popping up.  Except for the messageboxes the features seem to work fine.  See attached error log.

[RubberDuckLog vbe exception.txt](https://github.com/rubberduck-vba/Rubberduck/files/1804876/RubberDuckLog.vbe.exception.txt)

**Comment by bclothier** [comment]
Just to confirm the understanding -- you said it is randomly, so within a same session, the same project will sometime work, then other time it throws up that exception? 

Given the nature of exception, I would have expected the messagebox to consistently pop up all the time, not just sometimes.

**Comment by christophocles** [comment]
I was not really able to discern a pattern of when the errors occur. Within a particular session it might pop up only a couple of times, or it might pop up dozens of times. It is triggered by highlighting the name of a project or a module in the project explorer, but it does not always occur -only sometimes.  After a while it may just stop happening. It's pretty weird.

**Comment by bclothier** [comment]
Confirming.... You said the bad PC has Office 2013 and it used to have Access 2013 (and older).  Did you mean to say that you had Access runtime (or a standalone SKU) installed concurrently with an Office installation?

**Comment by bclothier** [comment]
Also, if you did have the runtime installed, does the Office come with Access included? There will be issues if a same version of Access runtime was installed to a PC containing Office that already has Access included.

**Comment by christophocles** [comment]
i currently have full standalone version of Access 2013. I used to have Access 2010 but this was uninstalled and replaced by the IT department several months ago. And I think Access 97 may have been present on this PC at one time but it has since been removed.  I can think of one custom access-based application I used in the past that may have included "runtime" components. So that's a possibility. Maybe I'll start with just reinstalling Access.

**Comment by bclothier** [comment]
And to be super duper sure, the Office does not come with Access included, right?

FWIW, if you did have runtime installed, they will have their own entries in the Programs and Features control panel, and after uninstalling you would most likely need to do a repair on both Office and then standalone Access. A PC should not have both runtime and full version of Access of same version installed together. It is only OK when the versions are different but even then, it has been known that multiple installation of Office can get you into a funky state like what you are seeing.

**Comment by carlos-quintero** [comment]
This error happens because MZ-Tools and RubberDuck are loaded in the same .NET AppDomain. (VBClassicExtensibility.* interop assembly belongs to MZ-Tools, Microsoft.Vbe.Interop.* interop assembly belongs to RubberDuck).

RubberDuck doesn't use a COM Shim to provide its own AppDomain, **although it should** because it will collide with any other .NET-based add-in for VBA ([Isolating .NET-based add-ins for the VBA editor with COM Shims](http://www.visualstudioextensibility.com/2016/07/11/isolating-net-based-add-ins-for-the-vba-editor-with-com-shims/)).

MZ-Tools does use a COM Shim to provide its own AppDomain, though. However, in your case it seems that is not using the COM shim, hence the collision of interop assemblies. You can verify the AppDomain on the MZ-Tools > About menu. If it says "MZToolsAppDomain" it is using the COM shim and no crash should occur. If it says "DefaultAppDomain" it is not using the COM shim and you will get a crash.

The only known reason for MZ-Tools not to use its COM Shim is if the setup detects AutoDesk installed, since it is not compatible with a COM Shim ([The strange case of MZ-Tools 8.0 crashing the VBA editor of Autodesk Inventor](http://www.visualstudioextensibility.com/2016/07/11/the-strange-case-of-mz-tools-8-0-crashing-the-vba-editor-of-autodesk-inventor/)).

Bottom line:
- Uninstall AutoDesk if not required and reinstall MZ-Tools. Check that the AppDomain is MZToolsAppDomain.
- If AutoDesk is already uninstalled, delete HKEY_LOCAL_MACHINE\Software\Autodesk (and/or HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Autodesk) and reinstall MZ-Tools. Check that the AppDomain is MZToolsAppDomain.
- If AutoDesk is required, then MZ-Tools, RubberDuck and any other .NET-based add-in cannot be loaded at the same time. You can load/unload using them the Add-In Manager.

**Comment by iamorangehat** [comment]
tried your program today and got the unexpected error, here's the bottom part of the log.
Note sure if it belongs in this thread or not. Thank you for your time.
Excel 2010 vba
2018-12-19 13:14:48.0371;ERROR-2.3.1.4308;Rubberduck.Parsing.VBA.ParseCoordinator;Unexpected exception thrown in parsing run. (thread 5).;System.Runtime.InteropServices.COMException (0x80029C4A): Error loading type library/DLL. (Exception from HRESULT: 0x80029C4A (TYPE_E_CANTLOADLIBRARY))
   at Microsoft.Vbe.Interop.Reference.get_IsBroken()
   at Rubberduck.VBEditor.SafeComWrappers.VBA.Reference.get_IsBroken() in C:\projects\rubberduck\Rubberduck.VBEditor.VBA\SafeComWrappers\VB\Reference.cs:line 67
   at Rubberduck.Parsing.VBA.ComReferenceLoading.COMReferenceSynchronizerBase.ProjectReferenceInfos(IVBProject project) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ComReferenceLoading\COMReferenceSynchronizerBase.cs:line 227
   at Rubberduck.Parsing.VBA.ComReferenceLoading.COMReferenceSynchronizerBase.ReferencedByProjectId() in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ComReferenceLoading\COMReferenceSynchronizerBase.cs:line 212
   at Rubberduck.Parsing.VBA.ComReferenceLoading.COMReferenceSynchronizerBase.SyncComReferences(CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ComReferenceLoading\COMReferenceSynchronizerBase.cs:line 84
   at Rubberduck.Parsing.VBA.ParsingStageService.SyncComReferences(CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParsingStageService.cs:line 92
   at Rubberduck.Parsing.VBA.ParseCoordinator.ExecuteCommonParseActivities(IReadOnlyCollection`1 toParse, IReadOnlyCollection`1 toReresolveReferencesInput, IReadOnlyCollection`1 newProjectIds, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 230
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAllInternal(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 497
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAll(Object requestor) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 406
2018-12-19 13:14:48.0501;INFO-2.3.1.4308;Rubberduck.Parsing.VBA.RubberduckParserState;RubberduckParserState (3) is invoking StateChanged (UnexpectedError);
2018-12-19 13:14:48.1321;DEBUG-2.3.1.4308;Rubberduck.Parsing.VBA.ParseCoordinator;Parsing run finished after 0.5485871s. (thread 5).;

**Comment by retailcoder** [comment]
@iamorangehat looks like your project references are broken? Please follow-up on #4663.

**Comment by agglomerator** [comment]
Revisiting this thread....

I have an Autodesk product, Fusion360, installed along with MZTools 8.0.0.1644 and Rubberduck 2.4.1.4627.  I keep getting the  ""InvalidCastException: Unable to cast object of type BClassicExtensibility.VBProjectClass' to type 'Microsoft.Vbe.Interop._VBProject" error even when I don't have Fusion360 running.  The solution stated was to unload/load MZTools and Rubberduck using the Add-in manager, but what does that mean?  I tried manually loading MZTools and Rubberduck using the AddIn manager and still got the "InvalidCastException" error.

Is there a way to have an Autodesk product AND MZTools and Rubberduck installed (not running) at the same time?  How do I do that when I want to run MzTools and Rubberduck ?  

Thanks for your help.

**Comment by carlos-quintero** [comment]
Hi,

I will provide technical information later, but first can you try this workaround?:

1) Using the Add-In Manager of the VBA editor:
- Mark Rubberduck to load on startup
- Mark MZ-Tools 8.0 **not** to load on startup
2) Close the Add-In Manager and the Autodesk product
3) Open the Autodesk product and the VBA editor. Rubberduck should load at this point, but not MZ-Tools.
4) Open the Add-In Manager, and load MZ-Tools 8.0 (but keep it marked not to load on startup).

At this point, neither Rubberduck nor MZ-Tools should crash ("InvalidCastException: Unable to cast object of type VBClassicExtensibility.VBProjectClass' to type 'Microsoft.Vbe.Interop._VBProject") when using features that use a project (ex: Task List in MZ-Tools, some other in Rubberduck).

Please confirm.

**Comment by bclothier** [comment]
Since the support itself seems to be addressed, we'll refer to issue #4041 for work on this area.

**Comment by agglomerator** [comment]
Hello Carlos - sorry I didn't get around to testing this.  This workaround
didn't solve the problem.  When I followed the steps outlined above, I got
the same "Interop" error any time I would click on an open worksheet, or
when I would click on the native VBE project explorer window.  I'm unable
to use Rubberduck any longer due to these errors.

If there is another workaround that can solve this issue, please let me
know.

Andre Kvaternik

**Comment by carlos-quintero** [comment]
@agglomerator I thought that the order of loading the add-ins was important but I was wrong.

Currently there is no workaround. If AutoDesk is required, then MZ-Tools, RubberDuck and any other .NET-based add-in cannot be loaded at the same time (InvalidCastException because they share the same default AppDomain but the interop types come from different assemblies). You can load/unload them using the Add-In Manager.

Ultimately it's a problem of AutoDesk, that crashes when a .NET-based add-in uses a COM Shim to use it's own AppDomain (rather than the shared default AppDomain) to avoid the InvalidCastException problem, but I doubt that AutoDesk is willing to solve the problem (they didn't answer my bug report when I reported it many months ago).

**Comment by bclothier** [comment]
I wonder if this [bug report](https://developercommunity.visualstudio.com/content/problem/422208/typeisequivalent-does-not-behave-according-to-the.html) is relevant? I believe that the implication are more widespread than just being not able to load an add-in.

**Comment by mansellan** [comment]
@carlos-quintero we're watching the development on .Net Core quite keenly at the moment, as version 3 will have support for WPF and COM interop. Microsoft have stated that AppDomains will never be supported in corefx. One idea that's been suggested is to write a shim that loads and forwards to the net core runtime - if we implement that, it will be interesting to see how that affects our isolation from other add-ins, I believe there is no precedence for this.
